### PR TITLE
automatically generate example output during testing

### PR DESCRIPTION
**Pull request by whoburg**
resolution to #30 
requires https://github.com/hoburg/gpkit/pull/513
@mjburton11 and @bqpd: ready for review.

### CODE REVIEW

**Comment by mjburton11** [comment]
Looks good to me.

**Comment by bqpd** [comment]
Yup, looks good.

File: gpfit/tests/t_examples.py [hunk 1]
```diff
@@ -48,9 +46,14 @@ def test_hoburgabbeel_ex6_1(self, example):
             setattr(TestExamples, name, old_test)  # move to a non-test fn
             delattr(TestExamples, fn)  # delete the old old_test
             new_name = "test_%s" % (name)
-            setattr(TestExamples, new_name, new_test(name, None))
+            setattr(TestExamples,
+                    new_name,
+                    logged_example_testcase(name=name,
+                                            imported=IMPORTED_EXAMPLES,
+                                            exampledir=EXAMPLE_DIR))
     TESTS.append(TestExamples)
```
Comment (whoburg) [inline gpfit/tests/t_examples.py]: @bqpd: the 16 lines above are replicated almost exactly in gpkit/tests/t_examples.py. The only difference is `logged_example_testcase` vs `new_test`.

Is there a better way to do this? somehow include more of this in the `logged_example_testcase` method over at GPkit? Make a new class that inherits from `unittest.TestCase` and handles all out fancy example logging tricks?

**Comment by whoburg** [comment]
This is waiting on https://github.com/hoburg/gpkit/pull/513.

File: gpfit/tests/t_examples.py (hunk 1 above)
Comment (bqpd) [inline gpfit/tests/t_examples.py]: how about 

```python
 TESTS = generate_example_tests(EXAMPLE_DIR, TestExamples, logged_example_testcase, SOLVERS)

def generate_example_tests(path, testclass, newtest_fn, solvers=None):
    if solvers is None:
        import gpkit
        solvers = gpkit.settings["installed_solvers"]
    if os.path.isdir(path):
        sys.path.insert(0, path)
        for fn in dir(testclass):
            if fn[:5] == "test_":
                 name = fn[5:]
                 old_test = getattr(testclass, fn)
                 setattr(testclass, name, old_test)  # move to a non-test fn
                 delattr(testclass, fn)  # delete the old old_test
                 for solver in solvers:
                     new_name = "test_%s_%s" % (name, solver)
                     setattr(testclass, new_name, newtest_fn(name, solver))
        TESTS.append(testclass)
```

Comment (whoburg) [inline gpfit/tests/t_examples.py]: looks good. Only change I'd make is that GPfit probably doesn't need to run for all solvers -- just the default.

Do you want to go ahead and push your suggestion to the `testhelpers` branch over at GPkit?

Docstring for `generate_example_tests` could be
```
"""
Mutate TestCase class so it behaves as described in TestExamples docstring

Arguments
---------
path (str): directory containing example modules to test
testclass (class): a class that inherits from `unittest.TestCase`
newtest_fn (function): method that returns new tests, e.g. logged_example_testcase
solvers (iterable): solvers to run for; or only for default if solvers is None
"""
```

Comment (bqpd) [inline gpfit/tests/t_examples.py]: Done! Check https://github.com/hoburg/gpkit/pull/513

**Comment by whoburg** [comment]
this is now failing due to `from gpkit.tests.helpers import generate_example_tests` -- that method was added to GPkit only recently -- do we need to update the gpkit Jenkins is using?

**Comment by whoburg** [comment]
merged #32 into this.

**Comment by galbramc** [comment]
tests are failing here as well, but it's a different failure

**Comment by whoburg** [comment]
interesting -- @bqpd we're getting a "Could not load settings file." in the console output for this -- more gpkit install issues.

**Comment by whoburg** [comment]
which gpkit does Jenkins install when testing? I don't think the gpkit changes that will make this work have been uploaded to PyPI yet.

**Comment by galbramc** [comment]
Jenkins creates a virtualenv and runs pip install gpkit. I was under the impression this is how you wanted it tested

**Comment by whoburg** [comment]
ok - that's exactly what's going on. Yes, you're right that this was how I wanted it tested. We'll have to wait on the next pip release of GPkit for this to work, then.

@bqpd: let's do another minor minor release -- I'm doing a tutorial tomorrow and would like folks installing to get the latest.

**Comment by bqpd** [comment]
Done...but the new version has a mysterious testing failure only in a new virtualenv. Debugging.

**Comment by bqpd** [comment]
Weird, there don't seem to have been any changes in that code path.

**Comment by galbramc** [comment]
retest this please

**Comment by galbramc** [comment]
retest this please

**Comment by galbramc** [comment]
retest this please

**Comment by galbramc** [comment]
retest this please

**Comment by galbramc** [comment]
retest this please

**Comment by galbramc** [comment]
retest this please

**Comment by galbramc** [comment]
retest this please

**Comment by galbramc** [comment]
retest this please

**Comment by galbramc** [comment]
retest this please

**Comment by galbramc** [comment]
retest this please

**Comment by whoburg** [comment]
more details on the "mysterious testing error"?

**Comment by whoburg** [comment]
also @galbramc: I noticed in the Jenkins console output the error below -- do the settings need to be updated for gpfit since we;re not doing separate `nounits` unit tests?

```
[xUnit] [INFO] - [JUnit] - No test report file(s) were found with the pattern 'test_reports/*.xml, test_reports_nounits/*.xml' relative to '/Users/jenkins/workspace/gpfit_PullRequest/buildnode/macys' for the testing framework 'JUnit'.  Did you enter a pattern relative to the correct directory?  Did you generate the result report(s) for 'JUnit'?
```

**Comment by galbramc** [comment]
Yes I'll update the settings to not look for no unit.

**Comment by galbramc** [comment]
retest this please

**Comment by galbramc** [comment]
retest this please

**Comment by galbramc** [comment]
retest this please

**Comment by galbramc** [comment]
This error seems to be sporadic, some times it passes, sometimes it fails:

`======================================================================
FAIL [0.000s]: test_rms_error (gpfit.tests.t_ex6_3.t_ex6_3_ISMA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home1/jenkins/workspace/gpfit_PullRequest/buildnode/reynolds/gpfit/tests/t_ex6_3.py", line 21, in test_rms_error
    self.assertTrue(self.rms_error < 5e-4)
AssertionError: False is not true
`

**Comment by galbramc** [comment]
retest this please

**Comment by whoburg** [comment]
Opened #33 to deal with the sporadic failure. I think this is ready to merge, let me now if anyone disagrees.

**Comment by whoburg** [comment]
Jenkins, do you agree? retest this please.

**Comment by galbramc** [comment]
retest this please

**Comment by bqpd** [comment]
Fixed! v0.3.4.9 pushed.

**Comment by galbramc** [comment]
@bqpd Except gpfit is now broken... 

https://acdl.mit.edu/csi/job/gpfit_Push/buildnode=reynolds/1/console

**Comment by bqpd** [comment]
Oops, sorry. I'll push the appropriate change to gpfit as well.

**Comment by bqpd** [comment]
That fixed it locally...

**Comment by galbramc** [comment]
Yeah it's fixed, there was a bug in the script. Could you add

[Build Status](https://acdl.mit.edu/csi/job/gpfit_Push/)

As the last line of README.md for gpfit

**Comment by bqpd** [comment]
Done!

**Pull request merged**